### PR TITLE
[JENKINS-58571] Check if name is editable on item before doing rename.

### PR DESCRIPTION
**Pull request by aHenryJard**
See [JENKINS-58571](https://issues.jenkins-ci.org/browse/JENKINS-58571).

### Proposed changelog entries

* Internal: Trying to rename items with non editable name is protected by exception raising.

### Submitter checklist

- [ X ] JIRA issue is well described
- [ X ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ X ] Appropriate autotests or explanation to why this change has no tests
- [ X ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/AbstractItem.java]: This looks very user unfriendly. Does the UI really allow users to get this far?

Comment (aHenryJard) [reply core/src/main/java/hudson/model/AbstractItem.java]: I am not sure that the UI permit users to do that, I just wanted to stop processing as soon as possible when it occurs. Anyway the same IllegalArgumentException will be raised just after on https://github.com/jenkinsci/jenkins/blob/217c02bece7afab237d8df669a6a21c80ac740a6/core/src/main/java/hudson/model/AbstractItem.java#L264

Maybe it's more user friendly to do something on https://github.com/jenkinsci/jenkins/blob/217c02bece7afab237d8df669a6a21c80ac740a6/core/src/main/java/hudson/model/AbstractItem.java#L277 and then return a FormValidationError.
@daniel-beck do you think it's better to remove this check or return a FormValidationError ?

Comment (aHenryJard) [reply core/src/main/java/hudson/model/AbstractItem.java]: I changed with a form validation error.

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: Maybe not needed? It is a dry-run, and it might still have use-cases when the name is not editable

File: core/src/test/java/hudson/model/AbstractItemTest.java
Comment (oleg-nenashev) [inline core/src/test/java/hudson/model/AbstractItemTest.java]:
```suggestion
    @Test
    @Issue("JENKINS-58571")
```

+ import, maybe

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractItem.java]: Please throw IOException which is a checked exception in API. Throwing an undocumented runtime exception may cause issues for API consumers

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Requesting changes, because undocumented runtime exceptions should not be thrown for `renameTo`

**Review by fcojfernandez — COMMENTED** [review]
Agree with @oleg-nenashev 's comment. It's better to throw an `IOException`

**Comment by oleg-nenashev** [comment]
Also created https://issues.jenkins-ci.org/browse/JENKINS-58681 which is somewhat related if there is a custom `ProjectNamingStrategy` enabled on the instance

**Comment by aHenryJard** [comment]
I looked at the tests failures they are all linked to the **MockFolder** that does not override the isNameEditable method to return true (and default is false on AbstractItem).

I see that this class is in https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/org/jvnet/hudson/test/MockFolder.java
I think that I need to do a PR in this project, but then the AbstractItem version in dependency is an old version without the isNameEditable method.

I look at it, and it come after 2.110 version of jenkins-core, that means that the version of jenkins-war should be updated to at least 2.110 (jenkins-core is a transitional dependency of jenkins-war) here https://github.com/jenkinsci/jenkins-test-harness/blob/04987e82891e65d27626b125e9ea0a2172c20ee3/pom.xml#L85

Do you think it's a good idea to increase this version ? And 2.110 is not the latest one, is it better to change latest instead ?

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (aHenryJard) [reply core/src/main/java/hudson/model/AbstractItem.java]: If I remove it, the Exception will be throws just some line after on https://github.com/jenkinsci/jenkins/blob/276d6899af202c288c9b5dabba05ed1d24cdfe20/core/src/main/java/hudson/model/AbstractItem.java#L264
I don't know how Exception are managed if it's not in FormValidation, I thought it was better for user experience. The renaming protection will still be there so I can remove this part.

**Comment by aHenryJard** [comment]
> Also created https://issues.jenkins-ci.org/browse/JENKINS-58681 which is somewhat related if there is a custom `ProjectNamingStrategy` enabled on the instance

I think that this issue is also related https://issues.jenkins-ci.org/browse/JENKINS-18649 because if there is a permission ITEM.rename that is created all this code will change.

It's on a comment on AbstractItem:
https://github.com/jenkinsci/jenkins/blob/276d6899af202c288c9b5dabba05ed1d24cdfe20/core/src/main/java/hudson/model/AbstractItem.java#L278

**Review by oleg-nenashev — APPROVED** [review]
Tests need to be fixed, but the code itself looks good to me

**Review by fcojfernandez — APPROVED** [review]

**Comment by aHenryJard** [comment]
Related PR to correct tests: https://github.com/jenkinsci/jenkins-test-harness/pull/159

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (jtnord) [reply core/src/main/java/hudson/model/AbstractItem.java]: > Maybe not needed? It is a dry-run, 

so you get no error from a check but then the operation will always fail, this is wrong on so many levels.

> and it might still have use-cases when the name is not editable

those UC are wrong, this is checking if you can rename this to "foobar" - you can not if the thing is not editable - if someone somehow calls this they should be told "No this is not a valid name becausr the thing can not be renamed".

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (jtnord) [reply core/src/main/java/hudson/model/AbstractItem.java]: > This looks very user unfriendly. Does the UI really allow users to get this far?

generally not, you will not see the link for a rename action.  It was just a belt and braes approach

**Comment by jglick** [comment]
> Related PR to correct tests

Prove that it works here: edit `test-pom/pom.xml` to request `2.55-rc1220.f5306ec770c9`.

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractItem.java]: Please revert unrelated whitespace or formatting changes, to keep the diff as short as possible. Assists review.

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractItem.java]: Not valid English. _Non-editable_ is, but perhaps you meant something like

```suggestion
            throw new IOException("Trying to rename an item that does not support this operation.");
```

?

Comment (jglick) [inline core/src/main/java/hudson/model/AbstractItem.java]: Should probably be throwing `Failure`, like `doConfirmRename` does. Or better yet, just move the call to `isNameEditable` into `doCheckNewName` where all the other checks are.

File: core/src/test/java/hudson/model/AbstractItemTest.java
Comment (jglick) [inline core/src/test/java/hudson/model/AbstractItemTest.java]: This is a pretty low-level test. That makes it fast, but on the other hand it is less clear that it really prevents the bug. A stronger test would actually make a real item (`createProject`) and try to use the `confirmRename` REST endpoint using `WebClient`. That however would require having an item type in the test classpath which is not a `Job` and which does not support renames, which would be trickier.

**Review by jglick — APPROVED** [review]
Some minor comments.

**Comment by jglick** [comment]
> edit `test-pom/pom.xml` to request `2.55-rc1220.f5306ec770c9`.

Now that would be whatever incremental release gets https://github.com/jenkinsci/jenkins-test-harness/pull/159/commits/14418be413ef04b187ab436aaa71168e515794ff. That did not get deployed yet since you need to bring the upstream PR up to date with its base branch, e.g.: `git pull origin master`

(See [JEP-305](https://jenkins.io/jep/305) for background.)

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (aHenryJard) [reply core/src/main/java/hudson/model/AbstractItem.java]: yes thanks it's better, I will take it.

Comment (aHenryJard) [reply core/src/main/java/hudson/model/AbstractItem.java]: I did that at the beginning, then I removed it because of @oleg-nenashev comment on it. I prefer to keep a protection here, because this is were action are really done, and we don't know how code will evolve in future or how class that inherit from AbstractItem will use it. I feel it's safer to check just before the real renaming.

For the form `Failure` in `doConfirmRename`, I can redo it again, I also think that it's better for user experience.

Comment (jglick) [reply core/src/main/java/hudson/model/AbstractItem.java]: Likely there is no UX here, since no GUI option to rename would have been offered to begin with if the item were not renamable.

No strong opinion about the placement or type of exception; an error is an error. I do not think anything external actually calls this method directly. Just check for CLI or REST usages.

File: core/src/test/java/hudson/model/AbstractItemTest.java
Comment (aHenryJard) [reply core/src/test/java/hudson/model/AbstractItemTest.java]: done

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (aHenryJard) [reply core/src/main/java/hudson/model/AbstractItem.java]: done.

**Comment by oleg-nenashev** [comment]
@aHenryJard please do not do force pushes while pull requests are in review. It makes hard to track changes during the review process, and maintainers do not get notifications about changes. We can always squash the PRs during the merge if needed.

File: test-pom/pom.xml
Comment (oleg-nenashev) [inline test-pom/pom.xml]:
```suggestion
      <version>2.55</version>
```

Released it few minutes ago

**Comment by oleg-nenashev** [comment]
Will merge tomorrow if no negative feedback.
Thanks @aHenryJard !

**Comment by daniel-beck** [comment]
This needs a followup change to fix Javadoc because it's now wrong.

`isNameEditable`:

> Controls whether the default rename action is available for this item.
> 
> `@return` whether {`@link` #name} can be modified by a user

`renameTo` is `protected`, so clearly not related to anything a user can do.

**Pull request merged**
